### PR TITLE
conflict detector(cdc): fix potential dead lock when a transaction generated multiple same hash

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/10334

### What is changed and how it works?

- Deduplicate the generated hash in a single transaction to prevent self cyclic dependency in the conflict detector.
- Refine the conflict detector, make it more readable

#### Tests 

 - Unit test

### Release note 

```release-note
Fix sink-to-mysql checkpoint-ts stuck issue in some rare cases.
```

### CODE REVIEW

File: pkg/causality/internal/node.go [hunk 1]
```diff
@@ -105,36 +105,43 @@ func (n *Node) NodeID() int64 {
...
 			return
 		}
 
 		if target.id == n.id {
 			panic("you cannot depend on yourself")
 		}
-		// Lock target and insert `n` into target.dependers.
+
+		// The target node might be removed or modified in other places, for exmaple
+		// after its corresponding transaction has been executed,
```
Comment (hicqu) [inline pkg/causality/internal/node.go:129]:
```suggestion
		// after its corresponding transaction has been executed.
```

File: pkg/causality/internal/node.go [hunk 2]
```diff
@@ -105,36 +105,43 @@ func (n *Node) NodeID() int64 {
...
 		target.mu.Lock()
 		defer target.mu.Unlock()
 
 		if target.assignedTo != unassigned {
 			// The target has already been assigned to a worker.
-			resolvedDependees = stdatomic.AddInt32(&n.resolvedDependees, 1)
+			// In this case, record the worker ID in `resolvedList`, and this node
+			// probably can be sent to the same worker and executed sequentially.
+			resolvedDependees = stdatomic.AddInt32(&n.resolvedDependencies, 1)
```
Comment (hicqu) [inline pkg/causality/internal/node.go:137]: You can search and replace all `Dependees` in the package.

File: cdc/sink/dmlsink/txn/event.go [hunk 3]
```diff
@@ -41,11 +41,40 @@ func (e *txnEvent) OnConflictResolved() {
...
+	}
+
+	// Sort hashes by `hash % numSlots` to avoid deadlock.
+	sort.Slice(hashes, func(i, j int) bool { return hashes[i]%numSlots < hashes[j]%numSlots })
+
+	// Dedup hashes
+	last := hashes[0]
+	j := 1
+	for i, hash := range hashes {
+		if i == 0 || hash == last {
```
Comment (CharlesCheung96) [inline cdc/sink/dmlsink/txn/event.go:68]:
```suggestion
		if hash == last {
```

File: pkg/causality/internal/node.go (hunk 2 above)
Comment (zhangjinpeng1987) [reply pkg/causality/internal/node.go]: done

File: cdc/sink/dmlsink/txn/event.go (hunk 3 above)
Comment (zhangjinpeng1987) [reply cdc/sink/dmlsink/txn/event.go]: I want to explicitly call out skip first element here.

**Review by asddongmen — APPROVED** [review]

**Review by sdojjy — APPROVED** [review]

**Comment by asddongmen** [comment]
Please merge the latest master branch into this pull request to ensure that the failed unit test can pass.

**Comment by asddongmen** [comment]
@3AceShowHand fixed the flaky unit test in https://github.com/pingcap/tiflow/pull/10327. You can update this PR when #10327 has been merged to pass the test.

**Pull request merged**